File: src/test/java/org/cloudfoundry/promregator/scanner/CachingTargetResolverTest.java
package org.cloudfoundry.promregator.scanner;

import java.util.LinkedList;
import java.util.List;

import org.cloudfoundry.promregator.JUnitTestUtils;
import org.cloudfoundry.promregator.lite.config.Target;
import org.cloudfoundry.promregator.scanner.MockedCachingTargetResolverSpringApplication.MockedTargetResolver;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = MockedCachingTargetResolverSpringApplication.class)
class CachingTargetResolverTest {
	@AfterAll
	static void cleanupEnvironment() {
		JUnitTestUtils.cleanUpAll();
	}

	@Autowired
	private TargetResolver targetResolver;
	
	@Autowired
	private CachingTargetResolver cachingTargetResolver;
	
	@AfterEach
	void resetFlags() {
		this.cachingTargetResolver.invalidateCache();
		
		( (MockedTargetResolver) targetResolver).resetRequestFlags();
		
	}
	@Test
	void testTwoPlainTargets() {
		List<Target> list = new LinkedList<>();
		list.add(MockedTargetResolver.Companion.getTarget1());
		list.add(MockedTargetResolver.Companion.getTarget2());
		
		List<ResolvedTarget> actualList = this.cachingTargetResolver.resolveTargets(list);
		
		MockedTargetResolver mtr = (MockedTargetResolver) targetResolver;
		Assertions.assertTrue(mtr.isRequestForTarget1());
		Assertions.assertTrue(mtr.isRequestForTarget2());
		Assertions.assertFalse(mtr.isRequestForTargetAllInSpace());
		Assertions.assertFalse(mtr.isRequestForTargetWithRegex());
		
		Assertions.assertEquals(2, actualList.size());
		
		
		boolean target1Found = false;
		boolean target2Found = false;
		for (ResolvedTarget rt : actualList) {
			if (rt == MockedTargetResolver.Companion.getRTarget1()) {
				target1Found = true;
			} else if (rt == MockedTargetResolver.Companion.getRTarget2()) {
				target2Found = true;
			} else {
				Assertions.fail("Unexpected target provided");
			}
		}
		Assertions.assertTrue(target1Found);
		Assertions.assertTrue(target2Found);
	}
	
	@Test
	void testTargetMissingApplicationName() {
		List<Target> list = new LinkedList<>();
		list.add(MockedTargetResolver.Companion.getTargetAllInSpace());
		
		List<ResolvedTarget> actualList = this.cachingTargetResolver.resolveTargets(list);
		
		MockedTargetResolver mtr = (MockedTargetResolver) targetResolver;
		Assertions.assertFalse(mtr.isRequestForTarget1());
		Assertions.assertFalse(mtr.isRequestForTarget2());
		Assertions.assertTrue(mtr.isRequestForTargetAllInSpace());
		Assertions.assertFalse(mtr.isRequestForTargetWithRegex());
		
		Assertions.assertEquals(2, actualList.size());
		
		boolean target1Found = false;
		boolean target2Found = false;
		for (ResolvedTarget rt : actualList) {
			if (rt == MockedTargetResolver.Companion.getRTarget1()) {
				target1Found = true;
			} else if (rt == MockedTargetResolver.Companion.getRTarget2()) {
				target2Found = true;
			} else {
				Assertions.fail("Unexpected target provided");
			}
		}
		Assertions.assertTrue(target1Found);
		Assertions.assertTrue(target2Found);
	}
	
	@Test
	void testRepeatedRequestIsCached() {
		List<Target> list = new LinkedList<>();
		list.add(MockedTargetResolver.Companion.getTarget1());
		
		// fill the cache
		List<ResolvedTarget> actualList = this.cachingTargetResolver.resolveTargets(list);
		
		MockedTargetResolver mtr = (MockedTargetResolver) targetResolver;
		Assertions.assertTrue(mtr.isRequestForTarget1());
		Assertions.assertFalse(mtr.isRequestForTarget2());
		Assertions.assertFalse(mtr.isRequestForTargetAllInSpace());
		Assertions.assertFalse(mtr.isRequestForTargetWithRegex());
		
		Assertions.assertEquals(1, actualList.size());
		
		ResolvedTarget rt = actualList.get(0);
		Assertions.assertEquals(MockedTargetResolver.Companion.getRTarget1(), rt);
		
		
		mtr.resetRequestFlags();
		
		actualList = this.cachingTargetResolver.resolveTargets(list);
		Assertions.assertFalse(mtr.isRequestForTarget1());
		Assertions.assertFalse(mtr.isRequestForTarget2());
		Assertions.assertFalse(mtr.isRequestForTargetAllInSpace());
		
		Assertions.assertEquals(1, actualList.size());
		
		rt = actualList.get(0);
		Assertions.assertEquals(MockedTargetResolver.Companion.getRTarget1(), rt);

	}
	
	@Test
	void testRepeatedRequestIsCachedAlsoSelectively() {
		List<Target> list = new LinkedList<>();
		list.add(MockedTargetResolver.Companion.getTarget1());
		
		// fill the cache
		List<ResolvedTarget> actualList = this.cachingTargetResolver.resolveTargets(list);
		
		MockedTargetResolver mtr = (MockedTargetResolver) targetResolver;
		Assertions.assertTrue(mtr.isRequestForTarget1());
		Assertions.assertFalse(mtr.isRequestForTarget2());
		Assertions.assertFalse(mtr.isRequestForTargetAllInSpace());
		Assertions.assertFalse(mtr.isRequestForTargetWithRegex());
		
		Assertions.assertEquals(1, actualList.size());
		
		ResolvedTarget rt = actualList.get(0);
		Assertions.assertEquals(MockedTargetResolver.Companion.getRTarget1(), rt);
		
		
		mtr.resetRequestFlags();
		
		list.add(MockedTargetResolver.Companion.getTarget2());
		
		actualList = this.cachingTargetResolver.resolveTargets(list);
		Assertions.assertFalse(mtr.isRequestForTarget1());
		Assertions.assertTrue(mtr.isRequestForTarget2());
		Assertions.assertFalse(mtr.isRequestForTargetAllInSpace());
		
		Assertions.assertEquals(2, actualList.size());
		
		boolean target1Found = false;
		boolean target2Found = false;
		for (ResolvedTarget rt2 : actualList) {
			if (rt2 == MockedTargetResolver.Companion.getRTarget1()) {
				target1Found = true;
			} else if (rt2 == MockedTargetResolver.Companion.getRTarget2()) {
				target2Found = true;
			} else {
				Assertions.fail("Unexpected target provided");
			}
		}
		Assertions.assertTrue(target1Found);
		Assertions.assertTrue(target2Found);
	}
	
	@Test
	void testTargetDuplicateRequestDistincts() {
		List<Target> list = new LinkedList<>();
		list.add(MockedTargetResolver.Companion.getTarget1());
		list.add(MockedTargetResolver.Companion.getTargetRegex());
		
		List<ResolvedTarget> actualList = this.cachingTargetResolver.resolveTargets(list);
		
		MockedTargetResolver mtr = (MockedTargetResolver) targetResolver;
		Assertions.assertTrue(mtr.isRequestForTarget1());
		Assertions.assertFalse(mtr.isRequestForTarget2());
		Assertions.assertFalse(mtr.isRequestForTargetAllInSpace());
		Assertions.assertTrue(mtr.isRequestForTargetWithRegex());
		
		Assertions.assertEquals(2, actualList.size());
		
		boolean target1Found = false;
		boolean target2Found = false;
		for (ResolvedTarget rt : actualList) {
			if (rt == MockedTargetResolver.Companion.getRTarget1()) {
				target1Found = true;
			} else if (rt == MockedTargetResolver.Companion.getRTarget2()) {
				target2Found = true;
			} else {
				Assertions.fail("Unexpected target provided");
			}
		}
		Assertions.assertTrue(target1Found);
		Assertions.assertTrue(target2Found);
	}
}

File: src/test/java/org/cloudfoundry/promregator/scanner/CachingTargetResolverTest.kt
package org.cloudfoundry.promregator.scanner

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.cloudfoundry.promregator.JUnitTestUtils
import org.cloudfoundry.promregator.config.Target
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.*

class CachingTargetResolverTest {
    private lateinit var targetResolver: TargetResolver
    private lateinit var cachingTargetResolver: CachingTargetResolver

    private val target1 = Target(orgName = "unittestorg", spaceName = "unittestspace", applicationName = "testapp", path = "path", protocol = "https")
    private val target2 = Target(orgName = "unittestorg", spaceName = "unittestspace", applicationName = "testapp2", path = "path", protocol = "https")
    private val targetRegex = Target(orgName = "unittestorg", spaceName = "unittestspace", applicationName = "testapp", path = "path", protocol = "https")

    private val rTarget1 = ResolvedTarget (target1)
    private val rTarget2 = ResolvedTarget (target2)

    @BeforeEach
    fun setup() {
        targetResolver = mockk()

        cachingTargetResolver = CachingTargetResolver(targetResolver, 300)
    }

    @Test
    fun `resolveTargets should pass through to parent resolver and be cached`() {
        every { targetResolver.resolveTargets(any())} returns listOf(rTarget1, rTarget2)

        val actualList = cachingTargetResolver.resolveTargets(listOf(target1, target2))

        assertThat(actualList).hasSize(2)

        assertThat(actualList).containsExactly(rTarget1, rTarget2)

        //Call a second time to trigger cache
        cachingTargetResolver.resolveTargets(listOf(target1, target2))

        verify(exactly = 1) { targetResolver.resolveTargets(any())} //Parent resolve should only have been called once
    }


    @Test
    fun `repeated calls only make requests for targets that are missing in the cache `() {
        val list= listOf(target1)

        every { targetResolver.resolveTargets(listOf(target1))} returns listOf(rTarget1)
        every { targetResolver.resolveTargets(listOf(target2))} returns listOf(rTarget2)

        // fill the cache
        val actualList = cachingTargetResolver.resolveTargets(list)
        assertThat(actualList).containsExactly(rTarget1)
        verify(exactly = 1) { targetResolver.resolveTargets(listOf(target1))}
        verify(exactly = 0) { targetResolver.resolveTargets(listOf(target2))} //Target 2 hasn't been called

        cachingTargetResolver.resolveTargets(listOf(target1, target2))
        verify(exactly = 1) { targetResolver.resolveTargets(listOf(target1))} //Taarget 1 is cached and isn't called aa second time
        verify(exactly = 1) { targetResolver.resolveTargets(listOf(target2))}
    }

    @Test
    fun `when duplicate targets are returned we only return the distinct ones`() {
        every { targetResolver.resolveTargets(any())} returns listOf(rTarget1, rTarget1, rTarget2)

        val list = listOf(target1, targetRegex)
        val actualList = cachingTargetResolver.resolveTargets(list)
        assertThat(actualList).hasSize(2)

        assertThat(actualList.filter { it == rTarget1 }).hasSize(1)
        assertThat(actualList.filter { it == rTarget2 }).hasSize(1)
    }

    companion object {
        @AfterAll
        fun cleanupEnvironment() {
            JUnitTestUtils.cleanUpAll()
        }
    }
}